Refactor commonalities between two approaches

awgu opened the pull request:
**Overview:**
This refactors some commonalities between the two approaches to overlapping DDP with ZeRO. This also partially addresses this comment: https://github.com/pytorch/pytorch/pull/62157#discussion_r680292827

**Test Plan:**
```
gpurun4 python test/distributed/optim/test_zero_redundancy_optimizer.py
```

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#62624 Refactor commonalities between two approaches**
* #62623 Add invariant check (bucket indices: 0, 1, ..., k-1)

Differential Revision: [D30058543](https://our.internmc.facebook.com/intern/diff/D30058543)

Commit:
Refactor commonalities between two approaches

[ghstack-poisoned]

Commit:
Update on "Refactor commonalities between two approaches"




[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30058543).

mrshenli commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -13,6 +13,106 @@
+    assert bucket_index in overlap_info.offsets, \
+        f"Bucket index {bucket_index} was not assigned to rank {rank}"
+    offset = overlap_info.offsets[bucket_index]
+    bucket_gradients = bucket.gradients()
```
nit: bucket_gradients seems just used once, looks like we don't need to create a var for it?

mrshenli commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -13,6 +13,106 @@
+    bucket: dist.GradBucket,
+    zero: ZeroRedundancyOptimizer,
+    rank: int,
+    rank_to_update: int,
```
`owner_rank`?

And also, would I be correct if I assume `owner_rank` is always the same as the `source_rank` argument in `_broadcast_bucket`, as the owner of those params should both update and broadcast. If yes, let's consolidate these two args to use the same name.

mrshenli reviewed (approved):
LGTM! Thanks for improving code readability.

Commit:
Update on "Refactor commonalities between two approaches"

**Overview:**
This refactors some commonalities between the two approaches to overlapping DDP with ZeRO. This also partially addresses this comment: https://github.com/pytorch/pytorch/pull/62157#discussion_r680292827

**Test Plan:**
```
gpurun4 python test/distributed/optim/test_zero_redundancy_optimizer.py
```


Differential Revision: [D30058543](https://our.internmc.facebook.com/intern/diff/D30058543)

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30058543).
